test/e2e/upgrade_cluster_test.go

```diff
@@ -0,0 +1,189 @@
+					tt.Logf("Found %d NodePools", len(nodepools.Items))
+					for i := range nodepools.Items {
+						tt.Logf("Found NodePool %s", nodepools.Items[i].Name)
+						nodePoolsMap[nodepools.Items[i].Name] = &nodepools.Items[i]
+					}
+				})
+				if !passed {
+					tt.Error("Failed to calculate HyperShift Operator upgrade invariants")
+					tt.Fail()
+					t.FailNow()
+				}
+				tt.Run("Upgrade HyperShift Operator", func(tt *testing.T) {
+					tt.Logf("Upgrading HyperShift Operator to image %s", hyperShiftOperatorLatestImage)
+					globalOpts.HyperShiftOperatorLatestImage = hyperShiftOperatorLatestImage
+					// Note that we're replacing previous HO installation cleanup function with the new one
+					cleanup, err = e2eutil.InstallHyperShiftOperatorFull(tt, ctx, globalOpts)
+					if err != nil {
+						tt.Error("Failed to upgrade HyperShift Operator")
+						tt.Fail()
+						t.FailNow()
+					}
+					// Set up the cleanup function on the outer test
+					t.Cleanup(func() {
+						once.Do(func() {
+							t.Logf("Cleaning up upgraded HyperShift Operator")
+							if err = cleanup(context.Background()); err != nil {
+								t.Logf("failed to cleanup HyperShift Operator: %v", err)
+							}
+						})
+					})
+				})
+				passed = tt.Run("Verify upgrade invariants", func(tt *testing.T) {
+					tt.Logf("Verifying upgrade invariants")
+					namespace := hostedCluster.Namespace
+					g := gomega.NewWithT(tt)
+
+					postUpgradeNodePools := &hyperv1.NodePoolList{}
+					if err = mgmtClient.List(ctx, postUpgradeNodePools, crclient.InNamespace(namespace)); err != nil {
+						tt.Fatalf("failed to list NodePools in namespace %s: %v", namespace, err)
+					}
+
+					g.Expect(postUpgradeNodePools.Items).ToNot(gomega.BeEmpty())
+					g.Expect(len(postUpgradeNodePools.Items)).To(gomega.BeEquivalentTo(len(nodePoolsMap)))
+
+					for _, nodePool := range postUpgradeNodePools.Items {
+						tt.Logf("Verifying NodePool %s", nodePool.Name)
+						preUpgradeNodePool, ok := nodePoolsMap[nodePool.Name]
+						g.Expect(ok).To(gomega.BeTrue())
+
+						tt.Logf("Generation: Want %d, Got %d", preUpgradeNodePool.Generation, nodePool.Generation)
+						tt.Logf("CurrentConfig: Want %s, Got %s", preUpgradeNodePool.Annotations[nodePoolAnnotationCurrentConfig], nodePool.Annotations[nodePoolAnnotationCurrentConfig])
+						tt.Logf("CurrentConfigVersion: Want %s, Got %s", preUpgradeNodePool.Annotations[nodePoolAnnotationCurrentConfigVersion], nodePool.Annotations[nodePoolAnnotationCurrentConfigVersion])
+
+						// Check if the node pool has been updated
+						g.Expect(nodePool.Generation).To(gomega.Equal(preUpgradeNodePool.Generation))
+						g.Expect(nodePool.Annotations[nodePoolAnnotationCurrentConfig]).To(gomega.Equal(preUpgradeNodePool.Annotations[nodePoolAnnotationCurrentConfig]))
+						g.Expect(nodePool.Annotations[nodePoolAnnotationCurrentConfigVersion]).To(gomega.Equal(preUpgradeNodePool.Annotations[nodePoolAnnotationCurrentConfigVersion]))
```

Comment:
can we please include a check for the underlying Deployment generation?

Reply:
I added the MD generation check, but I didn't use the `EnsureMachineDeploymentGeneration` as I wanted to surface more information in this test specifically like the MD name and log the generation.